src/main/java/org/springframework/integration/dsl/context/IntegrationFlowContext.java

```diff
@@ -75,69 +79,62 @@ public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
-	 * And preform its lifecycle start.
+	 * Associate provided {@link IntegrationFlow} with an {@link IntegrationFlowRegistrationBuilder}
+	 * for additional options and farther registration in the application context.
 	 * @param integrationFlow the {@link IntegrationFlow} to register
-	 * @return a generated id for the provided {@link IntegrationFlow}
+	 * @return the IntegrationFlowRegistrationBuilder associated with the provided {@link IntegrationFlow}
 	 */
-	public String register(IntegrationFlow integrationFlow) {
-		return register(integrationFlow, true);
+	public IntegrationFlowRegistrationBuilder registration(IntegrationFlow integrationFlow) {
+		return new IntegrationFlowRegistrationBuilder(integrationFlow);
 	}
 
-	/**
-	 * Register provided {@link IntegrationFlow} with generated {@code id}
-	 * in the application context as a bean together with all its dependant components.
-	 * And preform its lifecycle start if {@code autoStartup == true}.
-	 * @param integrationFlow the {@link IntegrationFlow} to register
-	 * @param autoStartup to start or not the {@link IntegrationFlow} automatically after registration
-	 * @return a generated id for the provided {@link IntegrationFlow}
-	 */
-	public String register(IntegrationFlow integrationFlow, boolean autoStartup) {
-		String id = generateBeanName(integrationFlow);
-		register(id, integrationFlow, autoStartup);
-		return id;
+	private void register(IntegrationFlowRegistrationBuilder builder) {
+		IntegrationFlow integrationFlow = builder.integrationFlowRegistration.getIntegrationFlow();
+		String flowId = builder.integrationFlowRegistration.getId();
+		if (flowId == null) {
+			flowId = generateBeanName(integrationFlow, null);
+			builder.id(flowId);
+		}
+		IntegrationFlow theFlow = (IntegrationFlow) registerBean(integrationFlow, flowId, null);
+		builder.integrationFlowRegistration.setIntegrationFlow(theFlow);
+
+		for (Map.Entry<Object, String> entry : builder.additionalBeans.entrySet()) {
+			registerBean(entry.getKey(), entry.getValue(), flowId);
+		}
+
+		if (builder.autoStartup) {
+			builder.integrationFlowRegistration.start();
+		}
+		this.registry.put(flowId, builder.integrationFlowRegistration);
 	}
 
-	/**
-	 * Register provided {@link IntegrationFlow} under provided {@code id}
-	 * in the application context as a bean together with all its dependant components.
-	 * And preform its lifecycle start.
-	 * @param flowId the bean name to register for
-	 * @param integrationFlow the {@link IntegrationFlow} to register
-	 */
-	public void register(String flowId, IntegrationFlow integrationFlow) {
-		register(flowId, integrationFlow, true);
+	private Object registerBean(Object bean, String beanName, String parentName) {
+		if (beanName == null) {
+			beanName = generateBeanName(bean, parentName);
+		}
+
+		bean = this.beanFactory.initializeBean(bean, beanName);
+		this.beanFactory.registerSingleton(beanName, bean);
+		this.autowiredAnnotationBeanPostProcessor.processInjection(bean);
```

Comment:
Need to do this before `initializeBean` since it calls init methods `afterPropertiesSet`, `@PostCostruct`.

Users will (reasonably) expect autowiring to happen before those calls.